### PR TITLE
4162: add version based optimistic row locking to tsk_aggregate_score…

### PR DESCRIPTION
**Pull request by raman-bt**
… table.

### CODE REVIEW

File: bindings/java/src/org/sleuthkit/datamodel/ScoringManager.java [hunk 1]
```diff
@@ -174,22 +182,38 @@ private void setAggregateScore(long objId, long dataSourceObjectId, Score score,
...
+
+					// register score change in the transaction.
+					transaction.registerScoreChange(new ScoreChange(objId, dataSourceObjectId, currentScore, resultScore));
+					return resultScore;
+				} else {
+					// return the current score
+					return currentScore;
+				}
+			} catch (SQLException ex) {
+				LOGGER.log(Level.WARNING, String.format("Error trying to update aggregate score for objId = %d, data source object id = %d.  Will try again. ", objId, dataSourceObjectId), ex);
```
Comment (APriestman) [inline bindings/java/src/org/sleuthkit/datamodel/ScoringManager.java:212]: You probably need to make a checkpoint and rollback the transaction here. I ran into errors with postgresql when I tried to continue using a transaction after an error. See https://www.progress.com/blogs/solution-to-current-transaction-is-aborted-error-for-postgresql-compatibility-with-odbc-jdbc

Comment (raman-bt) [reply bindings/java/src/org/sleuthkit/datamodel/ScoringManager.java:212]: Hmm.  The transaction is given to us by the calling client.  I don't think we can/should just roll it back  - it probably creates more trouble for the client than what we are trying to address.

@bcarrier, your thoughts?

Comment (raman-bt) [reply bindings/java/src/org/sleuthkit/datamodel/ScoringManager.java:212]: Ahh, I see what you mean - a savepoint for the transaction.  Yea - let me look into it.

Comment (APriestman) [reply bindings/java/src/org/sleuthkit/datamodel/ScoringManager.java:212]: I did it in createHost() in HostManager.java

File: bindings/java/src/org/sleuthkit/datamodel/ScoringManager.java [hunk 2]
```diff
@@ -114,47 +118,52 @@ private Score getAggregateScore(long objId, CaseDbConnection connection) throws
...
-	private void setAggregateScore(long objId, long dataSourceObjectId, Score score, CaseDbConnection connection) throws TskCoreException {
+	private void setAggregateScore(long objId, long dataSourceObjectId, Score score, int version, CaseDbConnection connection) throws TskCoreException, SQLException {
 
+		// insert or update score
 		String query = String.format("INSERT INTO tsk_aggregate_score (obj_id, data_source_obj_id, significance , confidence) VALUES (%d, %d, %d, %d)"
-				+ " ON CONFLICT (obj_id) DO UPDATE SET significance = %d, confidence = %d",
-				objId, dataSourceObjectId, score.getSignificance().getId(), score.getConfidence().getId(), score.getSignificance().getId(), score.getConfidence().getId() );
+				+ " ON CONFLICT (obj_id) DO UPDATE SET version = %d, significance = %d, confidence = %d WHERE version = %d",
+				objId, dataSourceObjectId, score.getSignificance().getId(), score.getConfidence().getId(), 
+				version + 1, score.getSignificance().getId(), score.getConfidence().getId(), version );
```
Comment (APriestman) [inline bindings/java/src/org/sleuthkit/datamodel/ScoringManager.java:155]: In postgresql I got this error:
org.postgresql.util.PSQLException: ERROR: column reference "version" is ambiguous

**Comment by bcarrier** [comment]
Instead of opportunistic locking, we are going to lock the table. We will keep this branch around though.

**Comment by bcarrier** [comment]
I was going to merge this into a branch for safe keeping, but it has merge conflicts that need to be resolved.

**Comment by raman-bt** [comment]
Its on my todo list to fix the conflicts.  I will do it right after finishing up account update.

**Pull request closed without merging**